src/disk_io_thread.cpp

```diff
@@ -1123,6 +1126,22 @@ namespace libtorrent
 
 		m_stats_counters.inc_stats_counter(counters::num_running_disk_jobs, -1);
 
+		mutex::scoped_lock l(m_cache_mutex);
+		if (m_cache_check_state == cache_check_idle)
+		{
+			m_cache_check_state = cache_check_active;
+			while (m_cache_check_state != cache_check_idle)
+			{
+				check_cache_level(l, completed_jobs);
+				--m_cache_check_state;
```

Comment:
would you mind adding an assert immediately following the call to check_cache_level() that the mutex is locked again? Just to make sure there's no future bug making check_cache_level() fail to re-acquire the lock.